<a href="https://colab.research.google.com/github/zzzer0-wav/myDTA_2026/blob/main/DTA_homework/numpy_homework_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🏠 Домашнє завдання — NumPy

**До тьюторіалу:** Pandas + Matplotlib: Аналіз даних з нуля

---
### Правила виконання
- Кожне завдання виконуй в окремій клітинці
- Не використовуй цикли там де можна обійтись операціями NumPy
- Після виконання — запусти клітинку і перевір що результат виглядає правильно
- Якщо застряг — поверніться до відповідного розділу тьюторіалу

**Оцінювання:** кожне завдання — 1 бал. Фінальний проєкт — 3 бали. Всього: **14 балів**

---

## Розділ 1 — Масиви ndarray та атрибути

> **Що повторюємо:** `np.array()`, `np.zeros()`, `np.ones()`, `np.arange()`, `np.linspace()`,
> атрибути `.shape`, `.ndim`, `.size`, `.dtype`

### ✏️ Завдання 1.1

У тебе є дані середньомісячних температур у Харкові за рік (°C):

```
-5, -4, 2, 10, 17, 21, 23, 22, 16, 9, 2, -3
```

1. Створи NumPy масив із цих значень і назви його `temps`
2. Виведи масив
3. Виведи: кількість вимірів (`ndim`), форму (`shape`), кількість елементів (`size`), тип даних (`dtype`)
4. Виведи мінімальну, максимальну температуру та середньорічну (округли до 1 знаку)

In [9]:
import numpy as np

# Твій код:
temps = np.array([-5, -4, 2, 10, 17, 21, 23, 22, 16, 9, 2, -3])

print(temps)
print('ndim:', temps.ndim)
print('shape:', temps.shape)
print('size:', temps.size)
print('dtype:', temps.dtype)

print('min:', min(temps))
print('max:', max(temps))
print('mean:', round(temps.mean(), 1))

[-5 -4  2 10 17 21 23 22 16  9  2 -3]
ndim: 1
shape: (12,)
size: 12
dtype: int64
min: -5
max: 23
mean: 9.2


### ✏️ Завдання 1.2

Створи таблицю планових продажів для 3 менеджерів на 4 квартали.

1. Створи матрицю `plan` розміром **3×4**, заповнену значенням `50` (тис. грн) — використай `np.full()`
2. Виведи матрицю
3. Виведи її форму та загальну кількість елементів
4. Створи другу матрицю `bonus` розміром **3×4** із нулів — менеджери ще не отримали бонусів
5. Виведи суму `plan + bonus` і переконайся що вона рівна `plan`

In [21]:
# Твій код:
plan = np.full((3, 4), 50)
print(plan)
print('shape: ', plan.shape)
print('size: ', plan.size)

bonus = np.zeros((3, 4))
print(bonus)

result = plan + bonus

print(result)

[[50 50 50 50]
 [50 50 50 50]
 [50 50 50 50]]
shape:  (3, 4)
size:  12
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
[[50. 50. 50. 50.]
 [50. 50. 50. 50.]
 [50. 50. 50. 50.]]


---
## Розділ 2 — Індексація та зрізи

> **Що повторюємо:** індексація 1D і 2D масивів, зрізи `[start:stop:step]`,
> вибір рядків та стовпців матриці

### ✏️ Завдання 2.1

Є масив денних продажів магазину за 2 тижні (14 днів):

```python
sales = np.array([320, 415, 290, 480, 350, 620, 580,
                  310, 440, 305, 510, 370, 650, 595])
```

Виконай такі вибірки **без циклів**, лише через зрізи:
1. Продажі першого тижня (дні 1–7)
2. Продажі другого тижня (дні 8–14)
3. Продажі у вихідні: 6-й, 7-й, 13-й, 14-й день — використай індексацію списком `[[5,6,12,13]]`
4. Продажі через день, починаючи з першого (непарні дні: 1, 3, 5...)
5. Масив у зворотному порядку

In [30]:
sales = np.array([320, 415, 290, 480, 350, 620, 580,
                  310, 440, 305, 510, 370, 650, 595])

# Твій код:
print(sales[0:7])
print(sales[7:14])
print(sales[[5,6,12,13]])
print(sales[0::2])
print(sales[::-1])

[320 415 290 480 350 620 580]
[310 440 305 510 370 650 595]
[620 580 650 595]
[320 290 350 580 440 510 650]
[595 650 370 510 305 440 310 580 620 350 480 290 415 320]


### ✏️ Завдання 2.2

Є матриця оцінок студентів: **4 студенти × 5 предметів**.

```python
grades = np.array([
    [85, 72, 91, 78, 88],   # Олена
    [63, 88, 74, 95, 70],   # Максим
    [90, 65, 82, 70, 94],   # Аня
    [75, 80, 68, 85, 79],   # Богдан
])
```

1. Виведи всі оцінки Олени (перший рядок)
2. Виведи оцінки всіх студентів з 3-го предмету (третій стовпець)
3. Виведи підматрицю: оцінки Максима та Ані з предметів 2–4 (рядки 1–2, стовпці 1–3)
4. Виведи оцінки останнього студента з останнього предмету

In [34]:
grades = np.array([
    [85, 72, 91, 78, 88],
    [63, 88, 74, 95, 70],
    [90, 65, 82, 70, 94],
    [75, 80, 68, 85, 79],
])

# Твій код:
print(grades[0])
print(grades[:,2])
print(grades[1:3, 1:4])
print(grades[-1, -1])

[85 72 91 78 88]
[91 74 82 68]
[[88 74 95]
 [65 82 70]]
79


---
## Розділ 3 — Математичні операції та векторизація

> **Що повторюємо:** арифметика з масивами, `np.sqrt()`, `np.round()`,
> агрегатні функції `.sum()`, `.mean()`, `.std()`, `.min()`, `.max()`, `np.median()`

### ✏️ Завдання 3.1

Є масив цін товарів у доларах. Виконай **без циклів**:

```python
prices_usd = np.array([15.99, 42.50, 8.00, 125.00, 33.75, 67.20, 9.99])
```

1. Переведи ціни в гривні (курс: **1 USD = 41.5 UAH**), результат округли до 2 знаків
2. Нарахуй ПДВ 20% — помнож гривневі ціни на 1.2, округли до 2 знаків
3. Знайди: мінімальну, максимальну, середню ціну з ПДВ
4. Знайди суму всіх товарів з ПДВ — це загальна вартість кошика
5. Виведи результати з поясненнями

In [55]:
prices_usd = np.array([15.99, 42.50, 8.00, 125.00, 33.75, 67.20, 9.99])

# Твій код:
prices_uah = np.round(prices_usd * 41.5, 2)
print(f'Ціни товарів у гривнях: {prices_uah}\n')

pdv20 = (prices_uah * 1.2)
print(f'+20% ПДВ {np.round(pdv20, 2)}\n')

print(f'Мінімальна ціна з ПДВ: {np.min(pdv20)} uah')
print(f'Максимальна ціна з ПДВ: {np.max(pdv20)} uah')
print(f'Сумма всіх товарів з ПДВ: {np.sum(pdv20)} uah')

Ціни товарів у гривнях: [ 663.58 1763.75  332.   5187.5  1400.62 2788.8   414.58]

+20% ПДВ [ 796.3  2116.5   398.4  6225.   1680.74 3346.56  497.5 ]

Мінімальна ціна з ПДВ: 398.4 uah
Максимальна ціна з ПДВ: 6225.0 uah
Сумма всіх товарів з ПДВ: 15060.996 uah


### ✏️ Завдання 3.2

Є дані про кількість товарів на складі та їх собівартість.
Виконай **без циклів**:

```python
quantity  = np.array([120, 45, 300, 18, 75, 200, 60])
cost_each = np.array([250, 1800, 45, 9500, 380, 120, 750])
names     = ['Кабель', 'Клавіатура', 'Зошит', 'Монітор',
             'Мишка', 'Ручка', 'Навушники']
```

1. Порахуй загальну вартість кожного товару на складі (`quantity * cost_each`)
2. Знайди загальну вартість всього складу
3. Знайди товар з найбільшою сумарною вартістю — використай `np.argmax()`
4. Знайди товар з найменшою сумарною вартістю — `np.argmin()`
5. Виведи середню вартість одного найменування на складі

In [74]:
quantity  = np.array([120, 45, 300, 18, 75, 200, 60])
cost_each = np.array([250, 1800, 45, 9500, 380, 120, 750])
names     = ['Кабель', 'Клавіатура', 'Зошит', 'Монітор',
             'Мишка', 'Ручка', 'Навушники']

# Твій код:
total = quantity * cost_each
print(f'Загальна вартість товарів на складі: {total}\n')
print(f'Загальна вартість всього складу: {np.sum(total)} грн\n')

max_idx = np.argmax(total)
min_idx = np.argmin(total)

print(f'Товар з найбільшою сумарною вартістю: {names[max_idx]} - {np.max(total)} грн')
print(f'Товар з найменшою сумарною вартістю: {names[min_idx]} - {np.min(total)} грн\n')
print(f'Середня вартість: {np.round(np.mean(total), 2)} грн')

Загальна вартість товарів на складі: [ 30000  81000  13500 171000  28500  24000  45000]

Загальна вартість всього складу: 393000 грн

Товар з найбільшою сумарною вартістю: Монітор - 171000 грн
Товар з найменшою сумарною вартістю: Зошит - 13500 грн

Середня вартість: 56142.86 грн


---
## Розділ 4 — Булеве маскування і фільтрація

> **Що повторюємо:** булеві маски, фільтрація `arr[mask]`,
> `np.where()`, `.sum()` для підрахунку True

### ✏️ Завдання 4.1

Є масив часу виконання замовлень у хвилинах (норма — до 30 хв):

```python
delivery_time = np.array([22, 35, 28, 41, 19, 55, 27, 33, 24, 48,
                           31, 26, 52, 18, 29, 38, 23, 44, 20, 36])
```

1. Знайди всі замовлення, виконані **вчасно** (≤ 30 хв)
2. Знайди всі **запізнілі** замовлення (> 30 хв)
3. Порахуй кількість вчасних і запізнілих
4. Знайди відсоток запізнілих замовлень (округли до 1 знаку)
5. Яка середня затримка серед запізнілих замовлень?

In [89]:
delivery_time = np.array([22, 35, 28, 41, 19, 55, 27, 33, 24, 48,
                           31, 26, 52, 18, 29, 38, 23, 44, 20, 36])

# Твій код:
intime = delivery_time[delivery_time <= 30]
late = delivery_time[delivery_time >= 30]
print('Замовлення, виконані вчасно: ', intime,'\n')
print('Запізнілі замовлення: ', late)

print(f'Кількість вчасних замовлень: {len(intime)}')
print(f'Кількість запізнілих замовлень: {len(late)}')
print(f'Відсоток запізнілих замовлень: {round(len(late) / len(delivery_time) * 100, 1)} %')
print(f'Середня затримка серед запізнілих замовлень: {np.mean(late)} хв')

Замовлення, виконані вчасно:  [22 28 19 27 24 26 18 29 23 20] 

Запізнілі замовлення:  [35 41 55 33 48 31 52 38 44 36]
Кількість вчасних замовлень: 10
Кількість запізнілих замовлень: 10
Відсоток запізнілих замовлень: 50.0 %
Середня затримка серед запізнілих замовлень: 41.3 хв


### ✏️ Завдання 4.2

Є масив щоденних продажів за місяць. Використай `np.where()`:

```python
np.random.seed(42)
daily = np.random.randint(100, 900, size=30)
```

1. Знайди дні з продажами **вище середнього** за місяць
2. Порахуй скільки таких днів
3. За допомогою `np.where()` створи масив `status`:
   - `'Гарний день'` — якщо продажі вище середнього
   - `'Звичайний день'` — якщо нижче або рівно
4. Виведи перші 10 днів: продажі та їх статус поряд

In [214]:
np.random.seed(42)
daily = np.random.randint(100, 900, size=30)

# Твій код:
print(daily,'\n')
avg_month = np.mean(daily)
print(f'                          Середнє за місяць: {round(avg_month, 2)}\n')

high_sales = daily[daily > avg_month]
days = np.where(daily > avg_month)[0] + 1

print(f'Дні з продажами більше за середнє за місяць: {days}')
print(f'                               Їхня виручка: {high_sales}')
print(f'                         Скільки таких днів: {len(days)}\n')

status = np.where(daily > avg_month, 'Гарний день', 'Звичайний день')

for i in range(10):
    print(f'День {i + 1}: продажі = {daily[i]}, статус = {status[i]}')


[202 535 370 206 171 800 120 714 221 566 314 430 558 187 472 199 763 230
 761 408 869 443 591 513 485 291 376 260 559 413] 

                          Середнє за місяць: 434.23

Дні з продажами більше за середнє за місяць: [ 2  6  8 10 13 15 17 19 21 22 23 24 25 29]
                               Їхня виручка: [535 800 714 566 558 472 763 761 869 443 591 513 485 559]
                         Скільки таких днів: 14

День 1: продажі = 202, статус = Звичайний день
День 2: продажі = 535, статус = Гарний день
День 3: продажі = 370, статус = Звичайний день
День 4: продажі = 206, статус = Звичайний день
День 5: продажі = 171, статус = Звичайний день
День 6: продажі = 800, статус = Гарний день
День 7: продажі = 120, статус = Звичайний день
День 8: продажі = 714, статус = Гарний день
День 9: продажі = 221, статус = Звичайний день
День 10: продажі = 566, статус = Гарний день


---
## Розділ 5 — reshape та сортування

> **Що повторюємо:** `.reshape()`, `.flatten()`, `np.sort()`, `np.argsort()`

### ✏️ Завдання 5.1

Є 1D масив з 24 показників продажів (по годинах за добу):

```python
hourly = np.array([12, 8, 5, 3, 4, 9, 18, 35, 52, 61, 58, 64,
                   70, 67, 55, 48, 53, 62, 74, 68, 45, 30, 22, 15])
```

1. Перетвори масив у матрицю **6×4** (6 блоків по 4 години) і виведи
2. Знайди суму продажів у кожному 4-годинному блоці (по рядках)
3. Відсортуй вихідний 1D масив за **зростанням** і виведи
4. Знайди **індекси** годин відсортованих за зростанням продажів — `np.argsort()`
5. Виведи топ-3 найпродуктивніших години (їхні номери, рахуючи з 0)

In [138]:
hourly = np.array([12, 8, 5, 3, 4, 9, 18, 35, 52, 61, 58, 64,
                   70, 67, 55, 48, 53, 62, 74, 68, 45, 30, 22, 15])

# Твій код:
fhours = hourly.reshape(6, 4)
print(fhours)

block_sums = np.sum(fhours, axis=1)
print('\nСуми рядків: ',block_sums)

sorted = np.sort(hourly)
print('\nВідсортований 1D масив: ',sorted,'')
print(f'Індекси відсортованих продажів: {np.argsort(hourly)}\n')

top3 = np.argsort(hourly)[::-1][:3]
print(f'Топ 3 найпродуктивніші години: {top3}')

[[12  8  5  3]
 [ 4  9 18 35]
 [52 61 58 64]
 [70 67 55 48]
 [53 62 74 68]
 [45 30 22 15]]

Суми рядків:  [ 28  66 235 240 257 112]

Відсортований 1D масив:  [ 3  4  5  8  9 12 15 18 22 30 35 45 48 52 53 55 58 61 62 64 67 68 70 74] 
Індекси відсортованих продажів: [ 3  4  2  1  5  0 23  6 22 21  7 20 15  8 16 14 10  9 17 11 13 19 12 18]

Топ 3 найпродуктивніші години: [18 12 19]


---
## Розділ 6 — Генерація випадкових чисел

> **Що повторюємо:** `np.random.seed()`, `np.random.randint()`,
> `np.random.normal()`, `np.random.choice()`

### ✏️ Завдання 6.1

Змоделюй результати маркетингового опитування 50 клієнтів.

1. Встанови `np.random.seed(99)`
2. Згенеруй масив `ages` — вік 50 клієнтів: цілі числа від 18 до 65
3. Згенеруй масив `scores` — оцінки задоволеності (нормальний розподіл, середнє=7, σ=1.5), округли до 1 знаку
4. Обріж `scores` так щоб значення були від 1 до 10 — `np.clip(arr, 1, 10)`
5. Виведи:
   - середній вік клієнтів
   - середню оцінку задоволеності
   - кількість клієнтів з оцінкою >= 8 («задоволені»)
   - відсоток задоволених

In [159]:
# Твій код:
np.random.seed(99)
ages = np.random.randint(18, 65, size=50)
scores = np.round(np.random.normal(7, 1.5, size=20), 1)
scores = np.clip(scores, 1, 10)

print(f'Середній вік клієнтів: {np.mean(ages)} роки\n')
print(f'Середня оцінка задоволеності: {np.mean(scores)}')

glad = scores[scores >= 8]
print(f'Середня кількість задоволених клієнтів (8+): {len(glad)}')
print(scores)

print(f'Відсоток задоволених: {len(glad) / len(scores) * 100}%')

Середній вік клієнтів: 41.4 роки

Середня оцінка задоволеності: 6.5
Середня кількість задоволених клієнтів (8+): 2
[5.7 9.2 5.5 6.1 6.8 7.9 4.8 8.8 7.5 6.4 7.2 6.2 6.7 7.4 6.  4.7 7.1 5.
 5.4 5.6]
Відсоток задоволених: 10.0%


---
## 🏆 Фінальний проєкт — Аналіз даних магазину

> **Використовуємо все з тьюторіалу разом**

Ти аналітик даних невеликого магазину електроніки.
Тобі передали масив продажів за 8 тижнів (56 днів).

Виконай повний аналіз та дай відповіді на запитання керівника.

In [160]:
# Дані — не змінюй цю клітинку
np.random.seed(2024)
base = np.random.randint(200, 800, 56)

# Вихідні продаються краще
weekend_boost = np.array([1 if i % 7 in [5, 6] else 0 for i in range(56)])
sales_56 = (base + weekend_boost * np.random.randint(100, 300, 56)).astype(int)

print('Дані отримано. Масив sales_56 готовий до аналізу.')
print(f'Елементів: {sales_56.size}, форма: {sales_56.shape}')


Дані отримано. Масив sales_56 готовий до аналізу.
Елементів: 56, форма: (56,)


### Запитання 1 — Загальна статистика (Розділ 1, 3)
Виведи:
- Загальна виручка за 56 днів
- Середньоденна виручка (округли до 2 знаків)
- Мінімальний та максимальний денний продаж
- Стандартне відхилення (наскільки продажі «скачуть»)

In [173]:
# Твій код:
total = np.sum(sales_56)
print(f'Загальна виручка: {total}$\n')
avg_day = np.mean(sales_56)
print(f'Середньоденна виручка: {np.round(avg_day, 2)}$\n')
print(f'Мінімальний денний продаж: {np.min(sales_56)}$')
print(f'Максимальний денний продаж: {np.max(sales_56)}$\n')
print(f'Стандартне відхилення: {np.round(np.std(sales_56), 2)}$\n')


Загальна виручка: 31819$

Середньоденна виручка: 568.2$

Мінімальний денний продаж: 222$
Максимальний денний продаж: 1072$

Стандартне відхилення: 195.52$



### Запитання 2 — Тижневий аналіз (Розділ 5)
1. Перетвори `sales_56` у матрицю **8×7** (8 тижнів по 7 днів)
2. Знайди виручку кожного тижня (сума по рядках)
3. Знайди найкращий тиждень і виведи його номер (рахуючи з 1) та суму
4. Знайди середньоденну виручку по кожному дню тижня (середнє по стовпцях)
5. Виведи який день тижня в середньому найприбутковіший (0=Пн, 6=Нд)

In [201]:
# Твій код:
newsales = sales_56.reshape(8, 7)
print(newsales, '\n')
weeks_total = np.sum(newsales, axis=1)
print(f'Виручка кожного тижня: {weeks_total}\n')
print(f'Найкращий тиждень: {np.argmax(weeks_total) + 1} з сумою - {weeks_total[4]}$\n')

print(f'Середньоденна виручка по тижням: {np.round(np.mean(newsales, axis=1), 1)}')

avg_by_day = np.mean(newsales, axis=0)

best_day = np.argmax(avg_by_day)
print(f'Найприбутковіший день тижня в середньому: {best_day} - неділя')

[[ 706  739  748  646  383  556  777]
 [ 427  274  522  680  708  465  845]
 [ 447  402  551  567  723  404  469]
 [ 254  314  383  754  297  472 1072]
 [ 773  510  769  798  510  824  852]
 [ 336  297  591  637  296  735  408]
 [ 661  447  734  558  339  710  779]
 [ 559  661  229  222  454  872  673]] 

Виручка кожного тижня: [4555 3921 3563 3546 5036 3300 4228 3670]

Найкращий тиждень: 5 з сумою - 5036$

Середньоденна виручка по тижням: [650.7 560.1 509.  506.6 719.4 471.4 604.  524.3]
Найприбутковіший день тижня в середньому: 6 - неділя


### Запитання 3 — Фільтрація та маскування (Розділ 4)
1. Знайди кількість **«слабких» днів** — продажі нижче 400 грн
2. Знайди кількість **«сильних» днів** — продажі вище 700 грн
3. Яка середня виручка у слабкі дні і яка у сильні?
4. Створи масив `labels` за допомогою `np.where()`:
   - `'Слабкий'` якщо < 400
   - `'Сильний'` якщо > 700
   - `'Нормальний'` в решті випадків (підказка: вкладений `np.where`)
5. Виведи перші 14 днів: продажі та їх мітки

In [220]:
# Твій код:
week_days = sales_56[sales_56 < 400]
strong_days = sales_56[sales_56 > 700]
print(f'Кількість слабких днів = {len(week_days)}')
print(f'Кількість сильних днів = {len(strong_days)}\n')
print(f'Середня виручка у слабкі дні: {np.round(np.mean(week_days), 2)}$')
print(f'Середня виручка у сильні дні: {np.round(np.mean(strong_days), 2)}$\n')
labels = np.where(sales_56 < 400, 'Слабкий', np.where(sales_56 > 700, 'Сильний', 'Нормальний'))

for i in range(14):
    print(f'День: {i + 1}, виручка: {sales_56[i]}, статус = {labels[i]}')


Кількість слабких днів = 12
Кількість сильних днів = 19

Середня виручка у слабкі дні: 302.0$
Середня виручка у сильні дні: 785.16$

День: 1, виручка: 706, статус = Сильний
День: 2, виручка: 739, статус = Сильний
День: 3, виручка: 748, статус = Сильний
День: 4, виручка: 646, статус = Нормальний
День: 5, виручка: 383, статус = Слабкий
День: 6, виручка: 556, статус = Нормальний
День: 7, виручка: 777, статус = Сильний
День: 8, виручка: 427, статус = Нормальний
День: 9, виручка: 274, статус = Слабкий
День: 10, виручка: 522, статус = Нормальний
День: 11, виручка: 680, статус = Нормальний
День: 12, виручка: 708, статус = Сильний
День: 13, виручка: 465, статус = Нормальний
День: 14, виручка: 845, статус = Сильний


**ВИСНОВОК**:
Продажі нестабільні (σ ≈ 195$), середій дохід - 568$. Є сильні колебання між днями (222-1072$), що може вказувати закономірність між слабкими днями, або на можливих нестабільних клієнтів.

Основний прибуток дають сильні дні (785$ / 302$ у слабких).
Кращий день тижня - неділя, кращий тиждень - 5.

Дохід залежить від пікових днів, є потенціал вирівнювання слабких перводів та посилення продажів у них (акції у слабки дні, реклама у дні просадок, таймінг цін [повисити у сильні дні та понизити трохи у слабкі] тощо)


---
## ✅ Підсумок

| Розділ | Завдання | Балів | Виконано |
|--------|----------|-------|----------|
| 1 — Масиви ndarray | 1.1, 1.2 | 2 | ☐ |
| 2 — Індексація та зрізи | 2.1, 2.2 | 2 | ☐ |
| 3 — Математичні операції | 3.1, 3.2 | 2 | ☐ |
| 4 — Булеве маскування | 4.1, 4.2 | 2 | ☐ |
| 5 — reshape та сортування | 5.1 | 1 | ☐ |
| 6 — Генерація випадкових | 6.1 | 1 | ☐ |
| 🏆 Фінальний проєкт | Питання 1–3 | 3 | ☐ |
| **Разом** | | **13** | |
